source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,386 @@
+      message_.memory_semantics_new_value_id());
+  if (!value_instruction) {
+    return false;
+  }
+  if (value_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+
+  // The value instruction must be an Integer.
+  if (ir_context->get_def_use_mgr()
+          ->GetDef(value_instruction->type_id())
+          ->opcode() != SpvOpTypeInt) {
+    return false;
+  }
+
+  // The size of the integer for value instruction must be equal to 32 bits.
+  auto value_instruction_int_width = ir_context->get_def_use_mgr()
+                                         ->GetDef(value_instruction->type_id())
+                                         ->GetSingleWordInOperand(0);
+
+  if (value_instruction_int_width != 32) {
+    return false;
+  }
+
+  // The first 5 bits of new memory semantics values must be suitable for needed
+  // instruction. The first 5 bits of new value must be larger than the first 5
+  // bits of old. Can't use Sequentially Consistent memory semantic if the
+  // memory model is Vulkan.
+  auto new_memory_sematics_value = value_instruction->GetSingleWordInOperand(0);
+  uint32_t old_memory_sematics_value = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    // Memory semantics true index for the OpMemoryBarrier equal 1.
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(1))
+            ->GetSingleWordInOperand(0);
+  } else {
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                GetMemorySemanticsOperandIndex(
+                    needed_atomic_instruction->opcode(),
+                        message_.memory_semantics_operand_index())))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto first_5bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto memory_model = static_cast<SpvMemoryModel>(
+      ir_context->module()->GetMemoryModel()->GetSingleWordInOperand(1));
+
+  if (!IsValidConversion(needed_atomic_instruction->opcode(),
+                         first_5bits_old_memory_semantics,
+                         first_5bits_new_memory_semantics, memory_model)) {
+    return false;
+  }
+
+  // The higher bits value of old and new memory semantics id must be equal.
+  auto higher_bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  auto higher_bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  if (higher_bits_new_memory_semantics != higher_bits_old_memory_semantics) {
+    return false;
+  }
```

Comment:
I suggest removing the casts in this case. The values may not end up as actual values of the enum, and we are not comparing against values of the enum, so there is no real advantage.

Reply:
Removed.